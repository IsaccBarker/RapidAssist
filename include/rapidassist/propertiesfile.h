/**********************************************************************************
 * MIT License
 * 
 * Copyright (c) 2018 Antoine Beauchamp
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *********************************************************************************/

#ifndef RA_PROPERTIESFILE_H
#define RA_PROPERTIESFILE_H

#include <string>
#include <map>

namespace ra { namespace filesystem {

  class PropertiesFile {
  public:
    PropertiesFile();
    virtual ~PropertiesFile();

    virtual bool load(const std::string & file_path);
    virtual bool save(const std::string & file_path);

    virtual bool clear();
    virtual bool hasKey(const std::string & key) const;
    virtual bool deleteKey(const std::string & key);
    virtual bool getValue(const std::string & key, std::string & value) const;
    virtual bool setValue(const std::string & key, const std::string & value);

  private:
    typedef std::map<std::string /*keyname*/, std::string /*value*/> PropertyMap;
    PropertyMap properties_;
  };

} //namespace filesystem
} //namespace ra

#endif //RA_PROPERTIESFILE_H
