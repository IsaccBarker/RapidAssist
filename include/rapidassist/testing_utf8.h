/**********************************************************************************
 * MIT License
 * 
 * Copyright (c) 2018 Antoine Beauchamp
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *********************************************************************************/

#ifndef RA_TESTING_UTF8_H
#define RA_TESTING_UTF8_H

#include <stdint.h>

#include "rapidassist/testing.h"

namespace ra { namespace testing {

#ifdef _WIN32

  //
  // Description:
  //  Returns true if the content of two files is identical.
  // Arguments:
  //   iFile1:          The path of the first  file.
  //   iFile2:          The path of the second file.
  //   oReason:         A textual reason why the files are not identical. Empty if files are identical.
  //   iMaxDifferences: The maximum number of textual differences.
  // 
  bool IsFileEqualsUtf8(const char* iFile1, const char* iFile2, std::string & oReason, size_t iMaxDifferences);

  inline bool IsFileEqualsUtf8(const char* iFile1, const char* iFile2) { std::string reason; return IsFileEqualsUtf8(iFile1, iFile2, reason, -1); }
  inline bool IsFileEqualsUtf8(const char* iFile1, const char* iFile2, std::string & oReason) { return IsFileEqualsUtf8(iFile1, iFile2, oReason, -1); }

  //
  // Description:
  //  Returns the location of all differences in two files.
  // Arguments:
  //   iFile1:          The path of the first  file.
  //   iFile2:          The path of the second file.
  //   oDifferences:    The list of all differences within both files. Empty if files are identical.
  //   iMaxDifferences: The maximum number of differences.
  // 
  bool GetFileDifferencesUtf8(const char* iFile1, const char* iFile2, std::vector<FileDiff> & oDifferences, size_t iMaxDifferences);

  //
  // Description:
  //  Returns true if the given text is found in a file.
  // Arguments:
  //   iFilename:   The path of the search file.
  //   iValue:      The search value.
  //   oLine:       The line number where iValue is found.
  //   oCharacter:  The character offset within the line.
  // 
  bool FindInFileUtf8(const char* iFilename, const char* iValue, int & oLine, int & oCharacter);

  //
  // Description:
  //  Creates a file of the given size. All bytes are sequential.
  // Arguments:
  //   iFilePath:   The path of the file.
  //   iSize:       The size in bytes of the file.
  // 
  bool CreateFileUtf8(const char * iFilePath, size_t iSize);

  //
  // Description:
  //  Creates text file.
  // Arguments:
  //   iFilePath:   The path of the file.
  // 
  bool CreateFileUtf8(const char * iFilePath);

  //
  // Description:
  //  Modify a given byte with the specified value.
  // Arguments:
  //   iFilePath: The path of the file.
  //   iOffset:   The offset of the modified byte.
  //   iValue:    The value of the replacement.
  // 
  void ChangeFileContentUtf8(const char * iFilePath, size_t iOffset, unsigned char iValue);

#elif __linux__

  //
  // Description:
  //  Returns true if the content of two files is identical.
  // Arguments:
  //   iFile1:          The path of the first  file.
  //   iFile2:          The path of the second file.
  //   oReason:         A textual reason why the files are not identical. Empty if files are identical.
  //   iMaxDifferences: The maximum number of textual differences.
  // 
  inline bool IsFileEqualsUtf8(const char* iFile1, const char* iFile2) { return IsFileEquals(iFile1, iFile2); }
  inline bool IsFileEqualsUtf8(const char* iFile1, const char* iFile2, std::string & oReason) { return IsFileEquals(iFile1, iFile2, oReason); }
  inline bool IsFileEqualsUtf8(const char* iFile1, const char* iFile2, std::string & oReason, size_t iMaxDifferences) { return IsFileEquals(iFile1, iFile2, oReason, iMaxDifferences); }

  //
  // Description:
  //  Returns the location of all differences in two files.
  // Arguments:
  //   iFile1:          The path of the first  file.
  //   iFile2:          The path of the second file.
  //   oDifferences:    The list of all differences within both files. Empty if files are identical.
  //   iMaxDifferences: The maximum number of differences.
  // 
  inline bool GetFileDifferencesUtf8(const char* iFile1, const char* iFile2, std::vector<FileDiff> & oDifferences, size_t iMaxDifferences) { return GetFileDifferences(iFile1, iFile2, oDifferences, iMaxDifferences); }

  //
  // Description:
  //  Returns true if the given text is found in a file.
  // Arguments:
  //   iFilename:   The path of the search file.
  //   iValue:      The search value.
  //   oLine:       The line number where iValue is found.
  //   oCharacter:  The character offset within the line.
  // 
  inline bool FindInFileUtf8(const char* iFilename, const char* iValue, int & oLine, int & oCharacter) { return FindInFile(iFilename, iValue, oLine, oCharacter); }

  //
  // Description:
  //  Creates a file of the given size. All bytes are sequential.
  // Arguments:
  //   iFilePath:   The path of the file.
  //   iSize:       The size in bytes of the file.
  // 
  inline bool CreateFileUtf8(const char * iFilePath, size_t iSize) { return CreateFile(iFilePath, iSize); }

  //
  // Description:
  //  Creates text file.
  // Arguments:
  //   iFilePath:   The path of the file.
  // 
  inline bool CreateFileUtf8(const char * iFilePath) { return CreateFile(iFilePath); }

  //
  // Description:
  //  Modify a given byte with the specified value.
  // Arguments:
  //   iFilePath: The path of the file.
  //   iOffset:   The offset of the modified byte.
  //   iValue:    The value of the replacement.
  // 
  inline void ChangeFileContentUtf8(const char * iFilePath, size_t iOffset, unsigned char iValue) { ChangeFileContent(iFilePath, iOffset, iValue); }


#endif

} //namespace testing
} //namespace ra

#endif //RA_TESTING_UTF8_H
