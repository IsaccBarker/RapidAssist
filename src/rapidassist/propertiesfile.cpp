/**********************************************************************************
 * MIT License
 * 
 * Copyright (c) 2018 Antoine Beauchamp
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *********************************************************************************/

#include "rapidassist/propertiesfile.h"

#include "rapidassist/strings.h"
#include "rapidassist/filesystem.h"
#include "rapidassist/testing.h"

#include <stdint.h>
#include <algorithm>    // std::min
#include <string.h>

namespace ra
{
  namespace filesystem
  {

    std::string processCppEscapeCharacter(const std::string & value) {
      //https://en.cppreference.com/w/cpp/language/escape
      //https://en.wikipedia.org/wiki/Escape_sequences_in_C
      struct EscapeSequence {
        const char * value;
        char replacement;
      };
      static const EscapeSequence sequences[] = {
        {"\\a", '\a'},  //beep
        {"\\b", '\b'},  //backspace
        {"\\e", 0x1b},  //esc
        {"\\f", '\f'},  //formfeed
        {"\\n", '\n'},  //newline
        {"\\r", '\r'},  //carriage return
        {"\\t", '\t'},  //tab
        {"\\v", '\v'},  //vertical tab
        {"\\\\", '\\'}, //backslash
        {"\\\'", '\''}, //single quotation mark
        {"\\\"", '\"'}, //double quotation mark
        {"\\?", '?'},   //question mark
        /*Unsupported escape sequences:
          \nnn
          \xhh
          \uhhhh
          \uhhhhhhhh
        */
      };
      static const size_t num_escape_sequences = sizeof(sequences) / sizeof(sequences[0]);

      std::string text;
      for (size_t i = 0; i < value.size(); i++) {
        const char * c = &value[i];

        //for each escape sequence
        bool isEscapedCharacter = false;
        for (size_t j = 0; isEscapedCharacter == false && j < num_escape_sequences; j++) {
          const EscapeSequence & seq = sequences[j];
          size_t seq_value_length = strlen(seq.value);
          if (strncmp(c, seq.value, seq_value_length) == 0) {
            //this is an escape sequence
            isEscapedCharacter = true;

            //add the escaped sequence instead of the character
            text.append(1, seq.replacement);
            i += (strlen(seq.value) - 1); //-1 to cancel this loop's increment
          }
        }

        if (!isEscapedCharacter) {
          text.append(1, *c);
        }
      }

      return text;
    }

    size_t findFirstUnescapedCharacter(const std::string & value, char c) {
      for (size_t i = 0; i < value.size(); i++) {
        char tmp = value[i];
        if (tmp == '\\') {
          i++; //also skip next escaped character
        }
        else if (tmp == c) {
          return i; //that's the character we are looking for
        }
      }

      return std::string::npos;
    }

    PropertiesFile::PropertiesFile() {
    }

    PropertiesFile::~PropertiesFile() {
    }

    bool PropertiesFile::load(const std::string & file_path) {
      if (file_path.empty())
        return false;

      if (!ra::filesystem::fileExists(file_path.c_str()))
        return false;

      clear();

      // TODO: process file open here
      ra::strings::StringVector lines;
      bool loaded = ra::testing::getTextFileContent(file_path.c_str(), lines);
      if (!loaded)
        return false;

      std::string multiline_key;
      std::string multiline_value; //accumulator for multiline values

      for (size_t i = 0; i < lines.size(); i++) {
        std::string line = lines[i];
        line = ra::strings::trim(line, ' ');
        if (line.empty())
          continue;
        else if (line[0] == '#' || line[0] == '!')
          continue; //this is a comment line

        //define placeholder for key and value
        std::string key;
        std::string value;

        //if we parsed a multiline key/value on previous line
        if (!multiline_key.empty()) {
          key = "";
          value = line;
        }
        else {
          //search for the first '=', ':' or unescaped ' ' character
          size_t split_pos = std::string::npos;
          size_t space_equal_pos = line.find(" = ");                       //for format "key = value"
          size_t space_colon_pos = line.find(" : ");                       //for format "key : value"
          size_t       equal_pos = findFirstUnescapedCharacter(line, '='); //for format "key=value"
          size_t       colon_pos = findFirstUnescapedCharacter(line, ':'); //for format "key:value"
          size_t       space_pos = findFirstUnescapedCharacter(line, ' '); //for format "key value"
          split_pos = std::min(split_pos, space_equal_pos);
          split_pos = std::min(split_pos, space_colon_pos);
          split_pos = std::min(split_pos, equal_pos);
          split_pos = std::min(split_pos, colon_pos);
          split_pos = std::min(split_pos, space_pos);
          if (split_pos != std::string::npos) {
            //compute split length
            size_t split_length = 1; //by default
            if (split_pos == space_equal_pos ||
              split_pos == space_colon_pos) {
              split_length = 3;
            }

            //split as "key" and "value"
            key = line.substr(0, split_pos);
            size_t value_length = line.size() - split_pos - split_length;
            value = line.substr(split_pos + split_length, value_length);
          }
          else
            continue; //no key-value split character in string

          //trim both to support the format "var=value" and "var = value"
          key = ra::strings::trimRight(key, ' ');
          value = ra::strings::trimLeft(value, ' ');
        }

        key = processCppEscapeCharacter(key);
        value = processCppEscapeCharacter(value);

        //process properties specific espcape characters on the key
        ra::strings::replace(key, "\\ ", " ");
        ra::strings::replace(key, "\\:", ":");
        ra::strings::replace(key, "\\=", "=");

        //is value continued on next line
        if (!value.empty() && value[value.size() - 1] == '\\') {
          //this key is a multiline value

          //remove last \ character
          value.erase(value.size() - 1, 1);

          //save the current key & value for the next line
          if (!key.empty())
            multiline_key = key;
          multiline_value += value;
        }
        else {
          //save the loaded key/value pair

          //are we dealing with a multiline value?
          if (!multiline_key.empty()) {
            multiline_value += value;
            this->setValue(multiline_key, multiline_value);
            multiline_key = "";
            multiline_value = "";
          }
          else {
            //single line key value
            this->setValue(key, value);
          }
        }

        int a = 0;
      }

      return true;
    }

    bool PropertiesFile::save(const std::string & file_path) {
      FILE * f = fopen(file_path.c_str(), "w");
      if (!f)
        return false;

      //for each properties
      for (PropertyMap::const_iterator propertyIt = mProperties.begin(); propertyIt != mProperties.end(); propertyIt++) {
        std::string key = propertyIt->first;
        const std::string & value = propertyIt->second;

        //remove key split characters
        ra::strings::replace(key, " ", "\\ ");
        ra::strings::replace(key, "=", "\\=");
        ra::strings::replace(key, ":", "\\:");

        fprintf(f, "%s = %s\n", key.c_str(), value.c_str());
      }

      fclose(f);
      return true;
    }

    bool PropertiesFile::clear() {
      mProperties.clear();
      return true;
    }

    bool PropertiesFile::hasKey(const std::string & key) const {
      PropertyMap::const_iterator propertyIt = mProperties.find(key);
      bool found = (propertyIt != mProperties.end());
      return found;
    }

    bool PropertiesFile::deleteKey(const std::string & key) {
      PropertyMap::iterator propertyIt = mProperties.find(key);
      bool found = (propertyIt != mProperties.end());
      if (!found)
        return true; //nothing to do

      mProperties.erase(propertyIt);
      return true;
    }

    bool PropertiesFile::getValue(const std::string & key, std::string & value) const {
      PropertyMap::const_iterator propertyIt = mProperties.find(key);
      bool found = (propertyIt != mProperties.end());
      if (!found)
        return false;

      value = propertyIt->second;
      return true;
    }

    bool PropertiesFile::setValue(const std::string & key, const std::string & value) {
      //overwrite previous property
      mProperties[key] = value;

      return true;
    }

  } //namespace filesystem
} //namespace ra
