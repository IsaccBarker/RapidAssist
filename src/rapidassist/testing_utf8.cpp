/**********************************************************************************
 * MIT License
 * 
 * Copyright (c) 2018 Antoine Beauchamp
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *********************************************************************************/

#include "rapidassist/testing_utf8.h"
#include "rapidassist/unicode.h"
#include "rapidassist/filesystem_utf8.h"

namespace ra { namespace testing {

#ifdef _WIN32 // UTF-8

  extern bool IsFileEquals(FILE* iFile1, FILE* iFile2, std::string & oReason, size_t iMaxDifferences);
  extern bool GetFileDifferences(FILE* iFile1, FILE* iFile2, std::vector<FileDiff> & oDifferences, size_t iMaxDifferences);

  //
  // Description:
  //  Wrapper class for FILE* instance.
  //  Which automatically closes the FILE* on wrapper's destruction
  // 
  class FileWrapperUtf8 {
  public:
    FileWrapperUtf8(const char * iPath, const char * iMode) {
      std::wstring pathW = ra::unicode::Utf8ToUnicode(iPath);
      std::wstring modeW = ra::unicode::Utf8ToUnicode(iMode);
      file_pointer_ = _wfopen(pathW.c_str(), modeW.c_str());
    }

    ~FileWrapperUtf8() {
      close();
    }

    bool isEOF() {
      if (file_pointer_ == NULL)
        return true;
      //http://www.cplusplus.com/reference/cstdio/feof/
      return (feof(file_pointer_) != 0);
    }

    void close() {
      if (file_pointer_) {
        fclose(file_pointer_);
        file_pointer_ = NULL;
      }
    }

    //members
    FILE * file_pointer_;
  };

  bool IsFileEqualsUtf8(const char* iFile1, const char* iFile2, std::string & oReason, size_t iMaxDifferences) {
    //Build basic message
    oReason.clear();
    oReason << "Comparing first file \"" << iFile1 << "\" with second file \"" << iFile2 << "\". ";

    FileWrapperUtf8 f1(iFile1, "rb");
    if (f1.file_pointer_ == NULL) {
      oReason << "First file is not found.";
      return false;
    }
    FileWrapperUtf8 f2(iFile2, "rb");
    if (f2.file_pointer_ == NULL) {
      oReason << "Second file is not found.";
      return false;
    }

    bool result = IsFileEquals(f1.file_pointer_, f2.file_pointer_, oReason, iMaxDifferences);
    return result;
  }

  bool GetFileDifferencesUtf8(const char* iFile1, const char* iFile2, std::vector<FileDiff> & oDifferences, size_t iMaxDifferences) {
    FileWrapperUtf8 f1(iFile1, "rb");
    if (f1.file_pointer_ == NULL)
      return false;
    FileWrapperUtf8 f2(iFile2, "rb");
    if (f2.file_pointer_ == NULL)
      return false;

    bool result = GetFileDifferences(f1.file_pointer_, f2.file_pointer_, oDifferences, iMaxDifferences);
    return result;
  }

  bool FindInFileUtf8(const char* iFilename, const char* iValue, int & oLine, int & oCharacter) {
    if (!ra::filesystem::FileExistsUtf8(iFilename))
      return false;

    oLine = -1;
    oCharacter = -1;

    ra::strings::StringVector lines;
    bool success = ra::filesystem::ReadTextFileUtf8(iFilename, lines);
    if (!success)
      return false;

    for (size_t i = 0; i < lines.size(); i++) {
      const std::string & line = lines[i];
      size_t position = line.find(iValue, 0);
      if (position != std::string::npos) {
        oLine = (int)i;
        oCharacter = (int)position;
        return true;
      }
    }

    return false;
  }

  bool CreateFileUtf8(const char * iFilePath, size_t iSize) {
    std::wstring pathW = ra::unicode::Utf8ToUnicode(iFilePath);
    FILE * f = _wfopen(pathW.c_str(), L"wb");
    if (!f)
      return false;

    for (size_t i = 0; i < iSize; i++) {
      unsigned int value = (i % 256);
      fwrite(&value, 1, 1, f);
    }
    fclose(f);
    return true;
  }

  bool CreateFileUtf8(const char * iFilePath) {
    std::wstring pathW = ra::unicode::Utf8ToUnicode(iFilePath);
    FILE * f = _wfopen(pathW.c_str(), L"w");
    if (f == NULL)
      return false;
    fputs("FOO!\n", f);
    fputs("&\n", f);
    fputs("BAR\n", f);
    fclose(f);

    return true;
  }

  void ChangeFileContentUtf8(const char * iFilePath, size_t iOffset, unsigned char iValue) {
    //read
    std::wstring pathW = ra::unicode::Utf8ToUnicode(iFilePath);
    FILE * f = _wfopen(pathW.c_str(), L"rb");
    if (!f)
      return;

    long size = ra::filesystem::GetFileSize(f);
    unsigned char * buffer = new unsigned char[size];
    if (!buffer)
      return;
    size_t byteRead = fread(buffer, 1, size, f);
    fclose(f);

    //modify
    if (iOffset < (size_t)size)
      buffer[iOffset] = iValue;

    //save
    f = _wfopen(pathW.c_str(), L"wb");
    if (!f)
      return;
    size_t byteWrite = fwrite(buffer, 1, size, f);
    fclose(f);
  }

#endif // UTF-8

} //namespace testing
} //namespace ra
