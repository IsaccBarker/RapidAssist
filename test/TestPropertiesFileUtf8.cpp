/**********************************************************************************
 * MIT License
 * 
 * Copyright (c) 2018 Antoine Beauchamp
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *********************************************************************************/

#include "TestPropertiesFileUtf8.h"

#include "rapidassist/propertiesfile.h"

#include "rapidassist/filesystem_utf8.h"
#include "rapidassist/testing_utf8.h"

namespace ra { namespace filesystem { namespace test
{
  //--------------------------------------------------------------------------------------------------
  void TestPropertiesFileUtf8::SetUp() {
  }
  //--------------------------------------------------------------------------------------------------
  void TestPropertiesFileUtf8::TearDown() {
  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestPropertiesFileUtf8, testLoadUtf8) {
    static const std::string path_separator = ra::filesystem::GetPathSeparatorStr();
    std::string test_name = ra::testing::GetTestQualifiedName();
    std::string file_source_path_nonutf8  = std::string("test_files") + path_separator + test_name + ".txt";
    std::string file_source_path_utf8     = std::string("test_files") + path_separator + test_name + ".psi_\xCE\xA8_psi.txt";

    bool copied = ra::filesystem::CopyFileUtf8(file_source_path_nonutf8, file_source_path_utf8);
    ASSERT_TRUE(copied);

    PropertiesFile s;
    ASSERT_TRUE(s.LoadUtf8(file_source_path_utf8));

    //assert expected values
    const std::string key01 = "website";
    const std::string key02 = "language";
    const std::string key03 = "message";
    const std::string key04 = "key with spaces";
    const std::string key05 = "tab";
    const std::string key06 = "path";
    const std::string expected01 = "https://en.wikipedia.org/";
    const std::string expected02 = "English";
    const std::string expected03 = "Welcome to Wikipedia!";
    const std::string expected04 = "This is the value that could be looked up with the key \"key with spaces\".";
    const std::string expected05 = "\\u0009"; //unicode characters not supported by PropertiesFile class. 
    const std::string expected06 = "c:\\wiki\\templates";
    std::string actual01;
    std::string actual02;
    std::string actual03;
    std::string actual04;
    std::string actual05;
    std::string actual06;
    ASSERT_TRUE(s.GetValue(key01, actual01));
    ASSERT_TRUE(s.GetValue(key02, actual02));
    ASSERT_TRUE(s.GetValue(key03, actual03));
    ASSERT_TRUE(s.GetValue(key04, actual04));
    ASSERT_TRUE(s.GetValue(key05, actual05));
    ASSERT_TRUE(s.GetValue(key06, actual06));

    ASSERT_EQ(expected01, actual01);
    ASSERT_EQ(expected02, actual02);
    ASSERT_EQ(expected03, actual03);
    ASSERT_EQ(expected04, actual04);
    ASSERT_EQ(expected05, actual05);
    ASSERT_EQ(expected06, actual06);
  }
  //--------------------------------------------------------------------------------------------------
  TEST_F(TestPropertiesFileUtf8, testSaveUtf8) {
    static const std::string path_separator = ra::filesystem::GetPathSeparatorStr();
    std::string test_name = ra::testing::GetTestQualifiedName();
    std::string file_expected_path = std::string("test_files") + path_separator + test_name + ".expected.txt";
    std::string file_actual_path_nonutf8  = std::string("test_files") + path_separator + test_name + ".actual.txt";
    std::string file_actual_path_utf8     = std::string("test_files") + path_separator + test_name + ".actual.psi_\xCE\xA8_psi.txt";

    PropertiesFile s;

    const std::string key01 = "website";
    const std::string key02 = "language";
    const std::string key03 = "key with spaces";
    const std::string key04 = "key=with=equals";
    const std::string key05 = "key:with:colons";
    const std::string value01 = "https://en.wikipedia.org/";
    const std::string value02 = "English";
    const std::string value03 = "this is the value of \"key with spaces\".";
    const std::string value04 = "this is the value of \"key=with=equals\".";
    const std::string value05 = "this is the value of \"key:with:colons\".";

    s.SetValue(key01, value01);
    s.SetValue(key02, value02);
    s.SetValue(key03, value03);
    s.SetValue(key04, value04);
    s.SetValue(key05, value05);

    ASSERT_TRUE(s.SaveUtf8(file_actual_path_utf8)) << "Failed saving properties to file '" << file_actual_path_utf8 << "'.";

    //assert that loading the saved file results in the same properties
    PropertiesFile s2;
    ASSERT_TRUE(s2.LoadUtf8(file_actual_path_utf8)) << "Failed loading property file '" << file_actual_path_utf8 << "'.";
    std::string read01;
    std::string read02;
    std::string read03;
    std::string read04;
    std::string read05;
    ASSERT_TRUE(s2.GetValue(key01, read01)) << "Failed reading the key '" << key01 << "' from file '" << file_actual_path_utf8 << "'.";
    ASSERT_TRUE(s2.GetValue(key02, read02)) << "Failed reading the key '" << key02 << "' from file '" << file_actual_path_utf8 << "'.";
    ASSERT_TRUE(s2.GetValue(key03, read03)) << "Failed reading the key '" << key03 << "' from file '" << file_actual_path_utf8 << "'.";
    ASSERT_TRUE(s2.GetValue(key04, read04)) << "Failed reading the key '" << key04 << "' from file '" << file_actual_path_utf8 << "'.";
    ASSERT_TRUE(s2.GetValue(key05, read05)) << "Failed reading the key '" << key05 << "' from file '" << file_actual_path_utf8 << "'.";
    ASSERT_EQ(value01, read01);
    ASSERT_EQ(value02, read02);
    ASSERT_EQ(value03, read03);
    ASSERT_EQ(value04, read04);
    ASSERT_EQ(value05, read05);
  }
  //--------------------------------------------------------------------------------------------------

} //namespace test
} //namespace filesystem
} //namespace ra
