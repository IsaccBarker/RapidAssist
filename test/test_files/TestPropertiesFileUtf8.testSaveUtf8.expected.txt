website = https://en.wikipedia.org/
language = English
key\ with\ spaces = this is the value of "key with spaces".
key\=with\=equals = this is the value of "key=with=equals".
key\:with\:colons = this is the value of "key:with:colons".
